Bring in Xilinx-AMD commits

bentheredonethat opened the pull request:
Bring in commits from Xilinx-AMD fork of OpenAMP System Reference Repo.

arnopo reviewed (changes requested):
I started to review this PR , but it is too big.  Please split in several commits.
Please also pay attention that examples should be as generic as possible, so put AMD specificity in `machine` when possible

bentheredonethat commented:
@arnopo  i have updated as follows
- removed all but AMD-Xilinx System-Device Tree commits 
- review-specific fixes

hopefully this will make it easier for you to review!

bentheredonethat commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -12,17 +12,23 @@
-#define LPRINTF(format, ...) printf(format, ##__VA_ARGS__)
-#define LPERROR(format, ...) LPRINTF("ERROR: " format, ##__VA_ARGS__)
+#define U_BOOT_EPT 1035
+#define U_BOOT_RPMSG_CHANNEL "U-Boot-rpmsg"
```
removed from this PR

bentheredonethat commented on examples/legacy_apps/examples/echo/freertos/main.c:
```diff
@@ -0,0 +1,107 @@
+						   NULL, NULL);
+		if (!rpdev) {
+			metal_err("Failed to create rpmsg virtio device.\r\n");
+			metal_err("RPU reboot is required to recover\r\n");
```
updated

bentheredonethat commented on examples/legacy_apps/examples/echo/freertos/main.c:
```diff
@@ -0,0 +1,107 @@
+	ret = platform_init(0, NULL, &platform);
+	if (ret) {
+		LPERROR("Failed to initialize platform.\r\n");
+		metal_err("RPU reboot is required to recover\r\n");
```
updated

bentheredonethat commented on examples/legacy_apps/examples/echo/freertos/main.c:
```diff
@@ -0,0 +1,107 @@
+
+	(void)(int *)unused_arg;
+
+	/* can't use metal_info, metal_log setup is in init_system */
```
removed from this PR

bentheredonethat commented on examples/legacy_apps/examples/echo/freertos/main.c:
```diff
@@ -0,0 +1,107 @@
+#include "task.h"
+
+#define LPRINTF(format, ...) xil_printf(format, ##__VA_ARGS__)
+#define LPERROR(fmt, ...) LPRINTF("ERROR: " fmt, ##__VA_ARGS__)
```
updated

bentheredonethat commented on examples/legacy_apps/examples/echo/CMakeLists.txt:
```diff
@@ -21,6 +26,11 @@ foreach (_app rpmsg-echo-ping rpmsg-echo)
   elseif (${_app} STREQUAL "rpmsg-echo")
     list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/rpmsg-echo.c")
+    if (${PROJECT_SYSTEM} STREQUAL "linux")
+      list (APPEND _sources "${CMAKE_CURRENT_SOURCE_DIR}/generic/main.c")
```
removed

bentheredonethat commented on examples/legacy_apps/examples/echo/freertos/main.c:
```diff
@@ -0,0 +1,107 @@
+#include "FreeRTOS.h"
+#include "task.h"
+
+#define LPRINTF(format, ...) xil_printf(format, ##__VA_ARGS__)
```
updated

tnmysh commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -115,23 +106,43 @@ int main(int argc, char *argv[])
 		LPERROR("Failed to initialize platform.\r\n");
-		ret = -1;
-	} else {
+		ML_ERR("Server reboot is required to recover\r\n");
```
Use metal_* macro.

bentheredonethat replied:
updated

tnmysh commented:
Squash this commit: [apps: rpmsg-echo: poll on vdev reset flag](https://github.com/OpenAMP/openamp-system-reference/pull/64/commits/b14116c6ae407677fed73c2d1255bf176bb8d2d4) into this one [apps: platform_info: add API for vdev reset flag](https://github.com/OpenAMP/openamp-system-reference/pull/64/commits/5932bb94bf1de5117c705cb9ab7a00dfb8dd08cf)

bentheredonethat commented:
> Squash this commit: [apps: rpmsg-echo: poll on vdev reset flag](https://github.com/OpenAMP/openamp-system-reference/pull/64/commits/b14116c6ae407677fed73c2d1255bf176bb8d2d4) into this one [apps: platform_info: add API for vdev reset flag](https://github.com/OpenAMP/openamp-system-reference/pull/64/commits/5932bb94bf1de5117c705cb9ab7a00dfb8dd08cf)

done

arnopo commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -115,23 +106,43 @@ int main(int argc, char *argv[])
+			 * Server behavior is undefined. It's better to wait in
+			 * an infinite loop instead
+			 */
+			while (1);
```
same issue here

bentheredonethat replied:
what here breaks? 

in our test env:
- for attach/detach  the while loop is ok because the RPU is re-sets up on re-attach

tnmysh replied:
@arnopo is that okay, if we can define return of main based on PROJECT_SYSTEM variable ?

#ifdef PROJECT_SYSTEM == LINUX
return <error code>;
#else
while (1);
#endif

arnopo replied:
That seems to me not enough generic.  
Behavior is platform dependent so I would be treated in platform_info.c ,  what about implementing in `cleanup_system` or a `system_shutdown `that would be empty by default?

tnmysh replied:
Yes that's a good idea. It will be re-factored accordingly.

bentheredonethat commented on examples/legacy_apps/machine/zynqmp_r5/platform_info.h:
```diff
@@ -64,6 +85,37 @@ extern "C" {
+
+#endif /* _AMD_GENERATED_ */
+
 struct remoteproc_priv {
```
Hi @arnopo  this value might be set by the amd generated header. so i kept it in the platform-info header.

bentheredonethat commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -115,23 +106,43 @@ int main(int argc, char *argv[])
+		 * Server behavior is undefined. It's better to wait in
+		 * an infinite loop instead
+		 */
+		while (1);
```
what here breaks? 

in our test env:
- for start / stop the while loop is ok because the RPU is rebooted via remoteproc

tnmysh commented on examples/legacy_apps/machine/zynqmp_r5/platform_info.h:
```diff
@@ -26,6 +29,24 @@ extern "C" {
+ *        OpenAMP YAML channel description. The generated symbols can be
+ *        changed by editing the OpenAMP YAML channel description.
+ */
+#include "amd_platform_info.h"
```
@arnopo we have not moved completely to S-DT flow yet. Once we move that, then we will always have amd_platform_info.h. For now for some platforms we still rely on hard-code symbols. So we need to keep above symbols along with amd_platofrm_info.h.

arnopo commented on examples/legacy_apps/examples/echo/rpmsg-echo.c:
```diff
@@ -115,23 +106,43 @@ int main(int argc, char *argv[])
 		LPERROR("Failed to initialize platform.\r\n");
-		ret = -1;
-	} else {
+		ML_ERR("Server reboot is required to recover\r\n");
```
One of the criteria for a PR  is that we are able to build the openamp lib, regardless of the commit we point to in the PR.

the build fails pointing on https://github.com/OpenAMP/openamp-system-reference/pull/64/commits/70598c4dd922534ec693f3455220682889571dde

tnmysh replied:
@bentheredonethat  just replace all the ML_* macros with metal_* macros from first PR. If it's already resolved, ignore this comment and resolve this conversation.

tnmysh commented:
@wmamills yes it's fine. It's not important, we will be building from downstream fork.

bentheredonethat commented:
@arnopo  sorry for long delay.. i have posted update.

bentheredonethat commented:
sorry again for the delay.. ping for review @wmamills  @arnopo @tnmysh

arnopo commented:
@bentheredonethat @tnmysh 

This PR is too big to be reviewed. Would it be possible to split it in severals PR? one PR per feature

bentheredonethat commented:
> @bentheredonethat @tnmysh
> 
> This PR is too big to be reviewed. Would it be possible to split it in severals PR? one PR per feature

sure how is the following to break it up:


> @bentheredonethat @tnmysh
> 
> This PR is too big to be reviewed. Would it be possible to split it in severals PR? one PR per feature

sure will break into few PRs.

bentheredonethat commented:
@arnopo  as a few different features are based on different things in early commits it is difficult to break up entirely. 

I have re-arranged and squashed commits to make it easier to review if that helps?

bentheredonethat commented:
@arnopo  will close open a smaller PR for you to review. thanks.
